x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentNodeServiceTests.java

```diff
@@ -356,6 +367,189 @@ public void testClusterChangedWithResetMode() {
         verifyNoMoreInteractions(deploymentManager, trainedModelAssignmentService);
     }
 
+    public void testClusterChanged_WhenAssigmentIsRoutedToShuttingDownNode_CallsStopAfterCompletingPendingWork()
+        throws InterruptedException {
+        final TrainedModelAssignmentNodeService trainedModelAssignmentNodeService = createService();
+        final DiscoveryNodes nodes = DiscoveryNodes.builder().localNodeId(NODE_ID).add(DiscoveryNodeUtils.create(NODE_ID, NODE_ID)).build();
+        String modelOne = "model-1";
+        String deploymentOne = "deployment-1";
+
+        ArgumentCaptor<TrainedModelDeploymentTask> stopParamsCapture = ArgumentCaptor.forClass(TrainedModelDeploymentTask.class);
+
+        doAnswer(invocationOnMock -> {
+            @SuppressWarnings({ "unchecked", "rawtypes" })
+            ActionListener<AcknowledgedResponse> listener = (ActionListener) invocationOnMock.getArguments()[1];
+            stopProcessCompletedLatch.countDown();
+            listener.onResponse(AcknowledgedResponse.TRUE);
+            return null;
+        }).when(trainedModelAssignmentService).updateModelAssignmentState(any(), any());
+
+        var taskParams = newParams(deploymentOne, modelOne);
+
+        ClusterChangedEvent event = new ClusterChangedEvent(
+            "testClusterChanged",
+            ClusterState.builder(new ClusterName("testClusterChanged"))
+                .nodes(nodes)
+                .metadata(
+                    Metadata.builder()
+                        .putCustom(
+                            TrainedModelAssignmentMetadata.NAME,
+                            TrainedModelAssignmentMetadata.Builder.empty()
+                                .addNewAssignment(
+                                    deploymentOne,
+                                    TrainedModelAssignment.Builder.empty(taskParams)
+                                        .addRoutingEntry(NODE_ID, new RoutingInfo(1, 1, RoutingState.STOPPING, ""))
+                                )
+                                .build()
+                        )
+                        .putCustom(NodesShutdownMetadata.TYPE, shutdownMetadata(NODE_ID))
+                        .build()
+                )
+                .build(),
+            ClusterState.EMPTY_STATE
+        );
+
+        trainedModelAssignmentNodeService.prepareModelToLoad(taskParams);
+        trainedModelAssignmentNodeService.clusterChanged(event);
+
+        if (stopProcessCompletedLatch.await(1, TimeUnit.MINUTES) == false) {
```

Comment:
Hmm, I thought this test would exercise the path through [gracefullyStopDeployment](https://github.com/elastic/elasticsearch/pull/98406/files#diff-0beebdd53ca3d25f539c171da17b5b58aeb6982ecb7ca34cf0e44acb94a586afR369) to eventually [stopDeploymentHelper](https://github.com/elastic/elasticsearch/pull/98406/files#diff-0beebdd53ca3d25f539c171da17b5b58aeb6982ecb7ca34cf0e44acb94a586afR521) which should execute on a separate thread right? Or is it still the main thread because we're using a test thread pool?